File: frontend/lib/core/models/transaction.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'category.dart';
import 'comment.dart';
import 'receipt.dart';

part 'transaction.freezed.dart';
part 'transaction.g.dart';

@freezed
class Transaction with _$Transaction {
  const factory Transaction({
    required String id,
    required String accountId,
    String? toAccountId,
    required String type,
    required double amount,
    required String currency,
    double? amountBase,
    required String date,
    String? description,
    String? categoryId,
    Category? category,
    @Default(false) bool aiCategorized,
    @Default(false) bool isRecurring,
    @Default([]) List<Comment> comments,
    @Default([]) List<Receipt> receipts,
    String? createdAt,
    String? updatedAt,
  }) = _Transaction;

  factory Transaction.fromJson(Map<String, dynamic> json) => _$TransactionFromJson(json);
}

@freezed
class PaginatedTransactions with _$PaginatedTransactions {
  const factory PaginatedTransactions({
    required List<Transaction> items,
    required int total,
    required int page,
    required int pageSize,
  }) = _PaginatedTransactions;

  factory PaginatedTransactions.fromJson(Map<String, dynamic> json) =>
      _$PaginatedTransactionsFromJson(json);
}

File: backend/app/schemas/transaction.py
import uuid
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, Field

from app.schemas.category import CategoryOut
from app.schemas.comment import CommentOut
from app.schemas.receipt import ReceiptOut


class TransactionBase(BaseModel):
    account_id: uuid.UUID
    to_account_id: Optional[uuid.UUID] = None
    type: str = Field(pattern="^(income|expense|transfer)$")
    amount: float
    currency: str
    date: date
    description: Optional[str] = None
    category_id: Optional[uuid.UUID] = None
    is_recurring: bool = False


class TransactionCreate(TransactionBase):
    pass


class TransactionUpdate(BaseModel):
    account_id: Optional[uuid.UUID] = None
    to_account_id: Optional[uuid.UUID] = None
    type: Optional[str] = None
    amount: Optional[float] = None
    currency: Optional[str] = None
    date: Optional[date] = None
    description: Optional[str] = None
    category_id: Optional[uuid.UUID] = None
    is_recurring: Optional[bool] = None


class TransactionOut(TransactionBase):
    id: uuid.UUID
    amount_base: Optional[float] = None
    ai_categorized: bool
    created_at: datetime
    updated_at: datetime
    category: Optional[CategoryOut] = None

    model_config = {"from_attributes": True}


class TransactionDetail(TransactionOut):
    comments: list[CommentOut] = []
    receipts: list[ReceiptOut] = []

    model_config = {"from_attributes": True}


class PaginatedTransactions(BaseModel):
    items: list[TransactionOut]
    total: int
    page: int
    page_size: int
